core/src/main/scala/org/apache/spark/util/ThreadUtils.scala

```diff
@@ -17,21 +17,104 @@
 
 package org.apache.spark.util
 
+import java.util
 import java.util.concurrent._
 import java.util.concurrent.locks.ReentrantLock
 
+import com.google.common.util.concurrent.{MoreExecutors, ThreadFactoryBuilder}
 import scala.concurrent.{Awaitable, ExecutionContext, ExecutionContextExecutor, Future}
 import scala.concurrent.duration.{Duration, FiniteDuration}
 import scala.language.higherKinds
 import scala.util.control.NonFatal
 
-import com.google.common.util.concurrent.ThreadFactoryBuilder
-
 import org.apache.spark.SparkException
 import org.apache.spark.rpc.RpcAbortException
 
 private[spark] object ThreadUtils {
 
+  object MDCAwareThreadPoolExecutor {
+    def newCachedThreadPool(threadFactory: ThreadFactory): ThreadPoolExecutor = {
+      // The values needs to be synced with [[Executors.newCachedThreadPool]]
+      new MDCAwareThreadPoolExecutor(
+        0,
+        Integer.MAX_VALUE,
+        60L,
+        TimeUnit.SECONDS,
+        new SynchronousQueue[Runnable],
+        threadFactory)
+    }
+
+    def newFixedThreadPool(nThreads: Int, threadFactory: ThreadFactory): ThreadPoolExecutor = {
+      // The values needs to be synced with [[Executors.newFixedThreadPool]]
+      new MDCAwareThreadPoolExecutor(
+        nThreads,
+        nThreads,
+        0L,
+        TimeUnit.MILLISECONDS,
+        new LinkedBlockingQueue[Runnable],
+        threadFactory)
+    }
+
+    def newSingleThreadExecutor(threadFactory: ThreadFactory): ThreadPoolExecutor = {
+      // The values needs to be synced with [[Executors.newSingleThreadExecutor]]
+      new MDCAwareThreadPoolExecutor(
```

Comment:
Executors.newSingleThreadExecutor returns a FinalizableDelegatedExecutorService. Compared with newFixedThreadPool, I think it is intentional as it states "Unlike the otherwise equivalent newFixedThreadPool(1, threadFactory) the returned executor is guaranteed not to be reconfigurable to use additional threads." 

Don't we need to follow it?

Reply:
Do you see anywhere in the code that we use that ability?

Reply:
@igreenfield, what if we use "that ability" in the future? Can you at least elaborate them in the comments if this is technically difficult?

Reply:
the class is package protected so I will need to copy that code. I will try to do it without copy and if it can't be done I will add a comment for that. Are you OK with that?

Reply:
Done